include/deal.II/numerics/vector_tools_evaluate.h

```diff
@@ -0,0 +1,252 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2021 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii_vector_tools_evaluation_h
+#define dealii_vector_tools_evaluation_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/mpi_remote_point_evaluation.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/fe/fe_point_evaluation.h>
+
+#include <map>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace VectorTools
+{
+  namespace EvaluationFlags
+  {
+    enum EvaluationFlags
+    {
+      avg    = 0,
+      max    = 1,
+      min    = 2,
+      insert = 3
+    };
+  }
```

Comment:
Before making this change, let's agree on how to deal with `gradients`, `hessians`! Is it an option that we have for each a separate function as stated a couple of minutes ago!? Or do we want to encrypt this in these "flags". An issue I seen in the latter is that we don't have a way to determine the return type of the function based on the flag so that the user might need to be explicit regarding the template argument!?